File: surface/widgets/compose/components/widgets/feedback/Dialog.kt
// ============================================================
// Clef Surface Compose Widget — Dialog
//
// Modal overlay that captures focus and blocks interaction with
// the underlying content until dismissed. Supports a title bar,
// description, arbitrary body content, and optional close-on-
// back-press behavior.
//
// Compose adaptation: Material 3 BasicAlertDialog (full custom
// layout) wrapped in a Surface with title, description, body
// content, and an optional close IconButton.
// See widget spec: repertoire/widgets/feedback/dialog.widget
// ============================================================

package clef.surface.compose.components.widgets.feedback

import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.Alignment
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.DialogProperties
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close

// --------------- Component ---------------

/**
 * Modal dialog overlay with title, description, and body content.
 *
 * @param open Whether the dialog is visible.
 * @param title Heading that labels the dialog.
 * @param description Supplementary text explaining the dialog purpose.
 * @param closeOnBackPress Whether pressing the system back button closes the dialog.
 * @param onClose Callback fired when the dialog is closed.
 * @param modifier Modifier applied to the dialog Surface.
 * @param content Composable content rendered inside the dialog body.
 */
@Composable
fun ClefDialog(
    open: Boolean = false,
    title: String? = null,
    description: String? = null,
    closeOnBackPress: Boolean = true,
    onClose: (() -> Unit)? = null,
    modifier: Modifier = Modifier,
    content: @Composable (ColumnScope.() -> Unit)? = null,
) {
    if (!open) return

    androidx.compose.ui.window.Dialog(
        onDismissRequest = {
            if (closeOnBackPress) {
                onClose?.invoke()
            }
        },
        properties = DialogProperties(
            dismissOnBackPress = closeOnBackPress,
            dismissOnClickOutside = closeOnBackPress,
        ),
    ) {
        Surface(
            modifier = modifier.fillMaxWidth(),
            shape = MaterialTheme.shapes.extraLarge,
            tonalElevation = 6.dp,
        ) {
            Column(
                modifier = Modifier.padding(24.dp),
            ) {
                // Title bar with optional close button
                if (title != null) {
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        modifier = Modifier.fillMaxWidth(),
                    ) {
                        Text(
                            text = title,
                            style = MaterialTheme.typography.headlineSmall,
                            modifier = Modifier.weight(1f),
                        )

                        if (closeOnBackPress) {
                            IconButton(
                                onClick = { onClose?.invoke() },
                                modifier = Modifier.size(24.dp),
                            ) {
                                Icon(
                                    imageVector = Icons.Filled.Close,
                                    contentDescription = "Close dialog",
                                    modifier = Modifier.size(20.dp),
                                )
                            }
                        }
                    }

                    Spacer(modifier = Modifier.height(8.dp))
                    HorizontalDivider()
                    Spacer(modifier = Modifier.height(16.dp))
                }

                // Description
                if (description != null) {
                    Text(
                        text = description,
                        style = MaterialTheme.typography.bodyMedium,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                    Spacer(modifier = Modifier.height(16.dp))
                }

                // Body content
                if (content != null) {
                    content()
                }
            }
        }
    }
}

File: surface/widgets/swiftui/components/widgets/feedback/Dialog.swift
// ============================================================
// Clef Surface SwiftUI Widget — Dialog
//
// Modal overlay that captures focus and blocks interaction with
// the underlying content until dismissed. Supports title bar,
// description, body content, and optional close button.
//
// Adapts the dialog.widget spec to SwiftUI rendering via sheet.
// ============================================================

import SwiftUI

// --------------- Component ---------------

/// Dialog view rendered as a modal sheet overlay.
///
/// - Parameters:
///   - open: Binding to whether the dialog is visible.
///   - title: Heading that labels the dialog.
///   - description: Supplementary text explaining the dialog purpose.
///   - closeOnBackPress: Whether the dialog can be dismissed externally.
///   - onClose: Callback fired when the dialog is closed.
///   - content: Composable content rendered inside the dialog body.
struct DialogView<Content: View>: View {
    @Binding var open: Bool
    var title: String? = nil
    var description: String? = nil
    var closeOnBackPress: Bool = true
    var onClose: (() -> Void)? = nil
    @ViewBuilder var content: Content

    var body: some View {
        EmptyView()
            .sheet(isPresented: $open, onDismiss: {
                onClose?()
            }) {
                VStack(alignment: .leading, spacing: 0) {
                    // Title bar
                    if let title = title {
                        HStack {
                            Text(title)
                                .font(.title3)
                                .fontWeight(.semibold)
                            Spacer()
                            if closeOnBackPress {
                                SwiftUI.Button(action: {
                                    open = false
                                    onClose?()
                                }) {
                                    Image(systemName: "xmark")
                                        .foregroundColor(.secondary)
                                }
                                .buttonStyle(.plain)
                                .accessibilityLabel("Close dialog")
                            }
                        }
                        .padding(.bottom, 8)

                        Divider()
                            .padding(.bottom, 16)
                    }

                    // Description
                    if let description = description {
                        Text(description)
                            .font(.body)
                            .foregroundColor(.secondary)
                            .padding(.bottom, 16)
                    }

                    // Body content
                    content
                }
                .padding(24)
                .interactiveDismissDisabled(!closeOnBackPress)
            }
    }
}
